File: src/main/java/com/fengsheng/skill/JingMeng.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.card.Card;
import com.fengsheng.phase.ReceivePhaseReceiverSkill;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.concurrent.TimeUnit;

/**
 * 程小蝶技能【惊梦】A：你接收黑色情报后，可以查看一名角色的手牌。
 */
public class JingMeng extends AbstractSkill implements TriggeredSkill {
    @Override
    public void init(Game g) {
        g.addListeningSkill(this);
    }

    @Override
    public SkillId getSkillId() {
        return SkillId.JING_MENG;
    }

    @Override
    public ResolveResult execute(Game g) {
        if (!(g.getFsm() instanceof ReceivePhaseReceiverSkill fsm) || fsm.inFrontOfWhom().findSkill(getSkillId()) == null)
            return null;
        if (fsm.inFrontOfWhom().getSkillUseCount(getSkillId()) > 0)
            return null;
        if (!fsm.messageCard().getColors().contains(Common.color.Black))
            return null;
        fsm.inFrontOfWhom().addSkillUseCount(getSkillId());
        return new ResolveResult(new executeJingMengA(fsm), true);
    }

    private record executeJingMengA(ReceivePhaseReceiverSkill fsm) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeJingMengA.class);

        @Override
        public ResolveResult resolve() {
            for (Player p : fsm.whoseTurn().getGame().getPlayers())
                p.notifyReceivePhase(fsm.whoseTurn(), fsm.inFrontOfWhom(), fsm.messageCard(), fsm.inFrontOfWhom(), 20);
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != fsm.inFrontOfWhom()) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (message instanceof Fengsheng.end_receive_phase_tos pb) {
                if (player instanceof HumanPlayer r && !r.checkSeq(pb.getSeq())) {
                    log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
                    return null;
                }
                player.incrSeq();
                return new ResolveResult(fsm, true);
            }
            if (!(message instanceof Role.skill_jing_meng_a_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Player r = fsm.inFrontOfWhom();
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (pb.getTargetPlayerId() < 0 || pb.getTargetPlayerId() >= g.getPlayers().length) {
                log.error("目标错误：" + pb.getTargetPlayerId());
                return null;
            }
            if (pb.getTargetPlayerId() == 0) {
                log.error("不能以自己为目标");
                return null;
            }
            Player target = g.getPlayers()[r.getAbstractLocation(pb.getTargetPlayerId())];
            if (!target.isAlive()) {
                log.error("目标已死亡");
                return null;
            }
            if (target.getCards().isEmpty()) {
                log.error("目标没有手牌");
                return null;
            }
            r.incrSeq();
            log.info(r + "发动了[惊梦]，查看了" + target + "的手牌");
            return new ResolveResult(new executeJingMengB(fsm, target), true);
        }
    }

    private record executeJingMengB(ReceivePhaseReceiverSkill fsm, Player target) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeJingMengB.class);

        @Override
        public ResolveResult resolve() {
            Player r = fsm.inFrontOfWhom();
            Game g = r.getGame();
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1) {
                    var builder = Role.skill_jing_meng_a_toc.newBuilder();
                    builder.setPlayerId(player1.getAlternativeLocation(r.location()));
                    builder.setTargetPlayerId(player1.getAlternativeLocation(target.location()));
                    builder.setWaitingSecond(20);
                    if (player1 == r) {
                        for (Card card : target.getCards().values())
                            builder.addCards(card.toPbCard());
                        final int seq2 = player1.getSeq();
                        builder.setSeq(seq2);
                        player1.setTimeout(GameExecutor.post(g, () -> {
                            if (player1.checkSeq(seq2)) {
                                Card card = null;
                                for (Card c : target.getCards().values()) {
                                    card = c;
                                    break;
                                }
                                assert card != null;
                                player1.getGame().tryContinueResolveProtocol(player1, Role.skill_jing_meng_b_tos.newBuilder()
                                        .setCardId(card.getId()).setSeq(seq2).build());
                            }
                        }, player1.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS));
                    }
                    player1.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer) {
                GameExecutor.post(g, () -> {
                    Card card = null;
                    for (Card c : target.getCards().values()) {
                        card = c;
                        break;
                    }
                    assert card != null;
                    r.getGame().tryContinueResolveProtocol(r, Role.skill_jing_meng_b_tos.newBuilder().setCardId(card.getId()).build());
                }, 2, TimeUnit.SECONDS);
            }
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != fsm.inFrontOfWhom()) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_jing_meng_b_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Player r = fsm.inFrontOfWhom();
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            Card card = target.findCard(pb.getCardId());
            if (card == null) {
                log.error("没有这张牌");
                return null;
            }
            r.incrSeq();
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1) {
                    var builder = Role.skill_jing_meng_b_toc.newBuilder();
                    builder.setPlayerId(player1.getAlternativeLocation(r.location()));
                    builder.setTargetPlayerId(player1.getAlternativeLocation(target.location()));
                    builder.setCard(card.toPbCard());
                    player1.send(builder.build());
                }
            }
            g.playerDiscardCard(target, card);
            return new ResolveResult(fsm, true);
        }
    }

    public static boolean ai(Fsm fsm0) {
        if (!(fsm0 instanceof executeJingMengA fsm))
            return false;
        Player p = fsm.fsm().inFrontOfWhom();
        Player target = fsm.fsm().whoseTurn();
        if (p == target || !target.isAlive() || target.getCards().isEmpty())
            return false;
        GameExecutor.post(p.getGame(), () -> p.getGame().tryContinueResolveProtocol(p, Role.skill_jing_meng_a_tos.newBuilder()
                .setTargetPlayerId(p.getAlternativeLocation(target.location())).build()), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/skill/JingMeng.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.RobotPlayer.Companion.bestCard
import com.fengsheng.protos.Fengsheng.end_receive_phase_tos
import com.fengsheng.protos.Role.skill_jing_meng_a_tos
import com.fengsheng.protos.Role.skill_jing_meng_b_tos
import com.fengsheng.protos.skillJingMengAToc
import com.fengsheng.protos.skillJingMengATos
import com.fengsheng.protos.skillJingMengBToc
import com.fengsheng.protos.skillJingMengBTos
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 程小蝶技能【惊梦】：你接收黑色情报后，可以查看一名角色的手牌。然后弃置其中一张牌，然后从中选择一张弃置。
 */
class JingMeng : TriggeredSkill {
    override val skillId = SkillId.JING_MENG

    override val isInitialSkill = true

    override fun execute(g: Game, askWhom: Player): ResolveResult? {
        val event = g.findEvent<ReceiveCardEvent>(this) { event ->
            askWhom === event.inFrontOfWhom || return@findEvent false
            g.players.any { it!!.alive && it.cards.isNotEmpty() }
            event.messageCard.isBlack()
        } ?: return null
        return ResolveResult(ExecuteJingMengA(g.fsm!!, event), true)
    }

    private data class ExecuteJingMengA(val fsm: Fsm, val event: ReceiveCardEvent) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            for (p in event.whoseTurn.game!!.players)
                p!!.notifyReceivePhase(event.whoseTurn, event.inFrontOfWhom, event.messageCard, event.inFrontOfWhom)
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== event.inFrontOfWhom) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message is end_receive_phase_tos) {
                if (player is HumanPlayer && !player.checkSeq(message.seq)) {
                    logger.error("操作太晚了, required Seq: ${player.seq}, actual Seq: ${message.seq}")
                    player.sendErrorMessage("操作太晚了")
                    return null
                }
                player.incrSeq()
                return ResolveResult(fsm, true)
            }
            if (message !is skill_jing_meng_a_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val r = event.inFrontOfWhom
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (message.targetPlayerId < 0 || message.targetPlayerId >= g.players.size) {
                logger.error("目标错误：${message.targetPlayerId}")
                player.sendErrorMessage("目标错误：${message.targetPlayerId}")
                return null
            }
            val target = g.players[r.getAbstractLocation(message.targetPlayerId)]!!
            if (!target.alive) {
                logger.error("目标已死亡")
                player.sendErrorMessage("目标已死亡")
                return null
            }
            if (target.cards.isEmpty()) {
                logger.error("目标没有手牌")
                player.sendErrorMessage("目标没有手牌")
                return null
            }
            r.incrSeq()
            logger.info("${r}发动了[惊梦]，查看了${target}的手牌")
            r.weiBiFailRate = 0
            return ResolveResult(ExecuteJingMengB(fsm, event, target), true)
        }
    }

    private data class ExecuteJingMengB(val fsm: Fsm, val event: ReceiveCardEvent, val target: Player) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            val r = event.inFrontOfWhom
            val g = r.game!!
            g.players.send { p ->
                skillJingMengAToc {
                    playerId = p.getAlternativeLocation(r.location)
                    targetPlayerId = p.getAlternativeLocation(target.location)
                    waitingSecond = Config.WaitSecond
                    if (p === r) {
                        target.cards.forEach { cards.add(it.toPbCard()) }
                        val seq2 = p.seq
                        seq = seq2
                        p.timeout = GameExecutor.post(g, {
                            if (p.checkSeq(seq2)) {
                                p.game!!.tryContinueResolveProtocol(p, skillJingMengBTos {
                                    cardId = target.cards.first().id
                                    seq = seq2
                                })
                            }
                        }, p.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) {
                GameExecutor.post(g, {
                    r.game!!.tryContinueResolveProtocol(r, skillJingMengBTos {
                        cardId = target.cards.bestCard(r.identity).id
                    })
                }, 3, TimeUnit.SECONDS)
            }
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== event.inFrontOfWhom) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_jing_meng_b_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val r = event.inFrontOfWhom
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            val card = target.findCard(message.cardId)
            if (card == null) {
                logger.error("没有这张牌")
                player.sendErrorMessage("没有这张牌")
                return null
            }
            r.incrSeq()
            logger.info("${r}弃掉了${target}的$card")
            g.players.send {
                skillJingMengBToc {
                    playerId = it.getAlternativeLocation(r.location)
                    targetPlayerId = it.getAlternativeLocation(target.location)
                    this.card = card.toPbCard()
                }
            }
            g.playerDiscardCard(target, card)
            g.addEvent(DiscardCardEvent(event.whoseTurn, target))
            return ResolveResult(fsm, true)
        }
    }

    companion object {
        fun ai(fsm0: Fsm): Boolean {
            if (fsm0 !is ExecuteJingMengA) return false
            val p = fsm0.event.inFrontOfWhom
            val target = p.game!!.players.filter {
                it!!.alive && p.isEnemy(it) && it.cards.isNotEmpty()
            }.randomOrNull() ?: return false
            GameExecutor.post(p.game!!, {
                p.game!!.tryContinueResolveProtocol(p, skillJingMengATos {
                    targetPlayerId = p.getAlternativeLocation(target.location)
                })
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}
